lib/wasix/src/runners/wasi_common.rs

```diff
@@ -257,6 +239,120 @@ fn create_dir_all(fs: &dyn FileSystem, path: &Path) -> Result<(), Error> {
     Ok(())
 }
 
+/// A directory that should be mapped from the host filesystem into a WASI
+/// instance (the "guest").
+#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
+pub struct MappedDirectory {
+    /// The absolute path for a directory on the host filesystem.
+    pub host: std::path::PathBuf,
+    /// The absolute path specifying where the host directory should be mounted
+    /// inside the guest.
+    pub guest: String,
+}
+
+#[derive(Debug, Clone)]
+pub struct MountedDirectory {
+    pub guest: String,
+    pub fs: Arc<dyn FileSystem + Send + Sync>,
+}
+
+impl From<MappedDirectory> for MountedDirectory {
+    fn from(value: MappedDirectory) -> Self {
+        let MappedDirectory { host, guest } = value;
+        let fs: Arc<dyn FileSystem + Send + Sync> =
+            Arc::new(HostFolderFileSystem::new(crate::default_fs_backing(), host));
+
+        MountedDirectory { guest, fs }
+    }
+}
+
+/// A [`FileSystem`] implementation that is scoped to a specific directory on
+/// the host.
+#[derive(Debug, Clone)]
+struct HostFolderFileSystem<F> {
+    root: PathBuf,
+    inner: F,
+}
+
+impl<F> HostFolderFileSystem<F> {
+    fn new(inner: F, root: PathBuf) -> Self {
+        HostFolderFileSystem { root, inner }
+    }
+
+    fn path(&self, path: &Path) -> Result<PathBuf, virtual_fs::FsError> {
+        let path = path.strip_prefix("/").unwrap_or(path);
```

Comment:
This allows operations outside of the directory with relative paths.

Note: I added something similar in Edge, which uses this code to check the path:

https://github.com/wasmerio/edge/blob/e6eb175163372dfd5261d7d3da281a1ad6bfb5bf/crates/edge/src/instance/volume/local.rs#L61

https://github.com/wasmerio/edge/blob/e6eb175163372dfd5261d7d3da281a1ad6bfb5bf/crates/edge/src/instance/volume/local.rs#L136